source/extensions/filters/udp/dns_filter/dns_parser.h

```diff
@@ -0,0 +1,189 @@
+                  const uint32_t ttl, Network::Address::InstanceConstSharedPtr ipaddr)
+      : BaseDnsRecord(query_name, rec_type, rec_class), ttl_(ttl), ip_addr_(ipaddr) {}
+
+  ~DnsAnswerRecord() override = default;
+  void serialize(Buffer::OwnedImpl& output) override { UNREFERENCED_PARAMETER(output); }
+
+  const uint32_t ttl_;
+  Network::Address::InstanceConstSharedPtr ip_addr_;
+};
+
+using DnsAnswerRecordPtr = std::unique_ptr<DnsAnswerRecord>;
+using DnsAnswerMap = std::unordered_multimap<std::string, DnsAnswerRecordPtr>;
+
+/**
+ * DnsQueryContext contains all the data associated with a query. The filter uses this object to
+ * generate a response and determine where it should be transmitted.
+ */
+class DnsQueryContext {
+public:
+  DnsQueryContext(Network::Address::InstanceConstSharedPtr local,
+                  Network::Address::InstanceConstSharedPtr peer)
+      : local_(std::move(local)), peer_(std::move(peer)), parse_status_(false), id_() {}
+  ~DnsQueryContext() = default;
+
+  Network::Address::InstanceConstSharedPtr local_;
+  Network::Address::InstanceConstSharedPtr peer_;
+
+  bool parse_status_;
+  uint16_t id_;
+  DnsQueryPtrVec queries_;
+  DnsAnswerMap answers_;
+};
+
+using DnsQueryContextPtr = std::unique_ptr<DnsQueryContext>;
+
+/**
+ * This class orchestrates parsing a DNS query and building the response to be sent to a client.
+ */
+class DnsMessageParser : public Logger::Loggable<Logger::Id::filter> {
+
+public:
+  enum DnsResponseCode { NoError, FormatError, ServerFailure, NameError, NotImplemented };
+
+  enum class DnsQueryParseState {
+    Init = 0,
+    Flags,      // 2 bytes
+    Questions,  // 2 bytes
+    Answers,    // 2 bytes
+    Authority,  // 2 bytes
+    Authority2, // 2 bytes
+    Finish
+  };
+
+  // These flags have been verified with dig. The flag order does not match the RFC, but takes byte
+  // ordering into account so that serialization does not need bitwise operations
+  PACKED_STRUCT(struct DnsHeaderFlags {
+    unsigned rcode : 4;  // return code
+    unsigned cd : 1;     // checking disabled
+    unsigned ad : 1;     // authenticated data
+    unsigned z : 1;      // z - bit (must be zero in queries per RFC1035)
+    unsigned ra : 1;     // recursion available
+    unsigned rd : 1;     // recursion desired
+    unsigned tc : 1;     // truncated response
+    unsigned aa : 1;     // authoritative answer
+    unsigned opcode : 4; // operation code
+    unsigned qr : 1;     // query or response
+  });
+
+  /**
+   * Structure representing the DNS header as it appears in a packet
+   * See https://www.ietf.org/rfc/rfc1035.txt for more details
+   */
+  PACKED_STRUCT(struct DnsHeader {
+    uint16_t id;
+    struct DnsHeaderFlags flags;
+    uint16_t questions;
+    uint16_t answers;
+    uint16_t authority_rrs;
+    uint16_t additional_rrs;
+  });
+
+  DnsMessageParser() = default;
+  ~DnsMessageParser() = default;
```

Comment:
nit: delete

Reply:
ping

Reply:
Assuming you mean to delete the destructor definition.